### PR TITLE
Clean shape inference

### PR DESCRIPTION
**Pull request by anderspapitto**

### CODE REVIEW

**Comment by anderspapitto** [comment]
[ do not review till https://github.com/onnx/onnx/pull/758 is pushed ]

File: onnx/defs/shape_inference.h [hunk 1]
```diff
@@ -44,8 +44,22 @@ inline bool hasExactlyNInputTypes(InferenceContext& ctx, int n, const std::strin
   return true;
 }
 
+inline bool hasNInputShapes(InferenceContext& ctx, int n, const std::string& opname) {
```
Comment (bddppq) [inline onnx/defs/shape_inference.h]: probably not worth adding a parameter "opname" just for error message (if you really want to show the opname, you can store the opname in the InferenceContext itself).

File: onnx/defs/shape_inference.h [hunk 2]
```diff
@@ -44,8 +44,22 @@ inline bool hasExactlyNInputTypes(InferenceContext& ctx, int n, const std::strin
...
+    if (!ctx.getInputType(i) || !ctx.getInputType(i)->has_shape()) {
+      return false;
+    }
+  }
+  return true;
+}
+
 inline void propagateElemTypeFromInputToOutput(InferenceContext& ctx, size_t inputIndex, size_t outputIndex) {
-  ctx.getOutputType(outputIndex)->set_elem_type(ctx.getInputType(inputIndex)->elem_type());
+  if (ctx.getInputType(inputIndex)) {
```
Comment (bddppq) [inline onnx/defs/shape_inference.h]: maybe assert inputIndex and outputIndex are not out of bound?

Comment (bddppq) [reply onnx/defs/shape_inference.h]: actually it might be better to assert inside getInputType and getOutputType

Comment (anderspapitto) [reply onnx/defs/shape_inference.h]: done, see implementation.h

File: onnx/shape_inference/implementation.h [hunk 3]
```diff
@@ -36,12 +36,18 @@ struct InferenceContextImpl : public InferenceContext {
     return allInputTypes_.size();
   }
   const TypeProto_Tensor* getInputType(size_t index) const override {
+    if (index >= allInputTypes_.size()) {
+      throw std::runtime_error("input " + std::to_string(index) + " is out of bounds");
```
Comment (bddppq) [inline onnx/shape_inference/implementation.h]: use ONNX_NAMESPACE::to_string

**Pull request merged**